Synching ALERT fork with main branch

mpaolone opened the pull request.

efuchey committed:
Changed the data bank to process the hits in AHDCEngine:
It should have been AHDC::adc (since corresponding HitReader also uses this bank), it was AHDC::tdc.

efuchey committed:
Merge branch 'JeffersonLab:development' into development

efuchey committed:
reversed the change in AHDCEngine.

efuchey committed:
Merge branch 'development' of github.com:efuchey/coatjava into development

efuchey committed:
Merge branch 'development' of github.com:efuchey/coatjava into development

efuchey committed:
Merge branch 'JeffersonLab:development' into development

efuchey committed:
Merge branch 'JeffersonLab:development' into development

efuchey committed:
Reverted (again???) the hit type to be read in AHDC reconstruction from TDC to ADC.

mpaolone committed:
changed AHDC bank back to ADC only

efuchey committed:
Merge branch 'JeffersonLab:development' into development

baltzell committed:
remove redundant checks

baltzell committed:
remove unused objects

baltzell committed:
cleanup

baltzell committed:
leave internal indexing at zero, output at 1

baltzell committed:
cleanup

baltzell committed:
remove redundant checks

baltzell committed:
leave internal indexing from zero, output from 1

baltzell committed:
simplicy logic

baltzell committed:
use s/l/c/o indexing starting from 1
(and stop changing wire number internally to start from zero)
(this requires removing a manual +1 offset for wire number in a perl script in GEMC)

mpaolone committed:
Merge pull request #1 from JeffersonLab/iss272-alert-indexing

Iss272 alert indexing

mpaolone committed:
merge post Nathan's indexing update

efuchey committed:
Merge branch 'JeffersonLab:development' into development

efuchey committed:
Merge branch 'JeffersonLab:development' into development

efuchey committed:
Merge pull request #1 from mpaolone/development

get latest changes

efuchey committed:
fixed a merge error that slipped through the cracks.

efuchey committed:
Merge branch 'JeffersonLab:development' into development

mpaolone committed:
Merge pull request #2 from efuchey/development

Fix of a merge error after commit f10c09e

efuchey committed:
Modified the reference layer index for the "OtherCluster" in the
PreCluster pairing routine in AHDC ClusterFinder

efuchey committed:
Modified the hardcoded lookup superlayer/layer indices to fill the per-layer list of hits in AHDC PreCluster
Modified the hardcoded lookup superlayer indices to obtain the geometry parameters to calculate x, y.

mpaolone committed:
Merge pull request #4 from efuchey/development

Modifications for debugging/validation of index convention change.

efuchey committed:
Declared upfront the hardcoded variables for cluster z calculation in Cluster.java

efuchey committed:
Merge branch 'JeffersonLab:development' into development

efuchey committed:
Added printouts to debug kalman filter.

efuchey committed:
Merge branch 'JeffersonLab:development' into development

efuchey committed:
Merge branch 'development' of github.com:efuchey/coatjava into development

efuchey committed:
Merge pull request #2 from mpaolone/development

get changes from Michael

efuchey committed:
Merge branch 'development' of github.com:efuchey/coatjava into development

efuchey committed:
Revert "Merge branch 'development' of github.com:efuchey/coatjava into development"

This reverts commit 62544e761c946da12186b3e6416eccebd4f6c915, reversing
changes made to 66afda9a051074af17212c28d1556886a3226d86.

efuchey committed:
Reapply "Merge branch 'development' of github.com:efuchey/coatjava into development"

This reverts commit abbb742512373cb2a4264b6282cc243a6a3bd0b3.

efuchey committed:
Adjusted the AHDC gas parameters in alert KalmanFilter.java

efuchey committed:
Merge branch 'JeffersonLab:development' into development

efuchey committed:
Merge branch 'JeffersonLab:development' into development

efuchey committed:
Completed commit 3420bb9 by modifying output index of atof paddle starting at 1 instead of 0

efuchey committed:
Merge branch 'development' of github.com:efuchey/coatjava into development

mpaolone committed:
Merge pull request #5 from efuchey/development

Completed commit 3420bb9

efuchey committed:
Added comments about units quantities.

efuchey committed:
Merge branch 'JeffersonLab:development' into development

efuchey committed:
Merge pull request #3 from mpaolone/development

Merge pull request #5 from efuchey/development

efuchey committed:
Merge branch 'development' of github.com:efuchey/coatjava into development

efuchey committed:
Merge branch 'JeffersonLab:development' into development

baltzell committed:
export BMT waveform

baltzell committed:
cleanup

efuchey committed:
Fixed the calculation of the hit wire position in Hit/Hit.java and KalmanFilter/Hit.java
Fixed the gas composition in MaterialMap;
Added some comments in KalmanFilter.java

baltzell committed:
add NamedEntry

efuchey committed:
Apparently, changing the wire indexing was not desirable, at least for Hit/Hit.java

baltzell committed:
refactor

baltzell committed:
add bank link

baltzell committed:
refactor

baltzell committed:
preserve index

baltzell committed:
tested, cleaned up

baltzell committed:
cleanup, add a mode-7

baltzell committed:
docs, examples

efuchey committed:
Verified and updated the calculations of the wire ends positions.
A few printouts are still left in.

mpaolone committed:
Merge branch 'JeffersonLab:development' into development

mpaolone committed:
Merge pull request #6 from JeffersonLab/baltzell-wf

The waveform proposal

efuchey committed:
Fixed B field direction in Kalman filter (anywhere else that it's wrong?)

Commit:
Implementation of a new mode of extraction

Creation of ModeAHDC.java

raffaelladevita committed:
bg-merger: using true order to inialize DetectorDescriptor

raffaelladevita committed:
one more switch between labbeledOrder and true order

raffaelladevita committed:
moved bg engine functionality to reuse bg events to common tools

Commit:
Update time calculations in ModeAHDC.java

ftouchte committed:
Merge pull request #1 from ftouchte/test/extractor

Update time calculations in ModeAHDC.java

raffaelladevita committed:
added functionality to merge multiple bg events

raffaelladevita committed:
added functionality to merge background for individual DC regions, plus bug fixed and some cleanup

Commit:
Reduce indentation

mpaolone committed:
Merge pull request #21 from ftouchte/development

Add a new mode of extraction for AHDC

efuchey committed:
* Fixed the magnetic field direction and the calculation of the wire end points;
* Adjusted the initial covariance matrix parameters;
* Added methods n KFitter and Hit to set better measurement and noise matrices for hits;

*** For debugging purposes only: ***
* Added a method to reconstruct that tracks gathering all hits, and sets the track parameters as follows:
   px = 150.0*sin(atan2(y_w0, x_w0)), py = 150.0*cos(atan2(y_w0, x_w0)) set with x_w0, y_w0 origin of first wire hit;
   x, y, z = 0, pz = 0;
   => only good for MC with no background;
* commented out the track finding method and using the method described above.

efuchey committed:
Merge branch 'development' of github.com:efuchey/coatjava into development

efuchey committed:
Added additional functions called "getMeasurementNoise_simple", "getVector_simple" for Hit, and "H_simple", "h_simple" for KFitter to return the preexisting matrix emasurements;
new matrix measurements have been implemented in the methods "getMeasurementNoise", "getVector", "H" and "h".
Added new memebers "adc" and "phi" for the hit for measured ADC and wire phi position at z = 0 respectively.

efuchey committed:
Merge pull request #4 from mpaolone/development

Getting changes from Michael

efuchey committed:
Merge branch 'development' of github.com:efuchey/coatjava into development

raffaelladevita commented on common-tools/clas-detector/src/main/java/org/jlab/detector/base/DetectorDescriptor.java:
```diff
@@ -64,6 +64,14 @@ public void setOrder(int order){
         }
     }
     
+    public int[] getCSC() {
```
Are these methods used anywhere? if not, they should be removed

baltzell replied:
Not sure it's used, but I wouldn't remove it just because of that

raffaelladevita commented on reconstruction/alert/src/main/java/org/jlab/rec/ahdc/Cluster/Cluster.java:
```diff
@@ -27,7 +27,9 @@ public Cluster(PreCluster precluster, PreCluster other_precluster) {
 		this._Radius = (precluster.get_Radius() + other_precluster.get_Radius()) / 2;
-		this._Z      = ((other_precluster.get_Phi() - precluster.get_Phi()) / (Math.toRadians(20) * Math.pow(-1, precluster.get_Super_layer()) - Math.toRadians(20) * Math.pow(-1, other_precluster.get_Super_layer()))) * 300 - 150;
+
+		this._Z      = ((other_precluster.get_Phi() - precluster.get_Phi()) / (Math.toRadians(20) * Math.pow(-1, precluster.get_Super_layer()-1) - Math.toRadians(20) * Math.pow(-1, other_precluster.get_Super_layer()-1))) * 300 - 150;
```
hardcoded numbers should be replaced with variables possibly defined from the database or in this case the geometry service

whit2333 replied:
Agree! Can we merge this now and create a new issue for this?

whit2333 replied:
Created issue.

raffaelladevita commented on reconstruction/alert/src/main/java/org/jlab/rec/ahdc/Cluster/ClusterFinder.java:
```diff
@@ -13,14 +13,16 @@ public class ClusterFinder {
 			ArrayList<PreCluster> possible_precluster_list = new ArrayList<>();
 
-			double phi_mean = precluster.get_Phi() + 0.1 * Math.pow(-1, precluster.get_Super_layer());
+			double phi_mean = precluster.get_Phi() + 0.1 * Math.pow(-1, precluster.get_Super_layer()-1);
```
same comment about hardcoded numbers

baltzell replied:
I think it's the "0.1" in question

whit2333 replied:
@raffaelladevita  and @baltzell can we create a new issue to circle back and fix this? We really need to get this behemoth merged so we can make progress (ideally with small, more manageable PRs).

raffaelladevita commented on reconstruction/alert/src/main/java/org/jlab/rec/ahdc/Hit/Hit.java:
```diff
@@ -31,31 +31,31 @@ private void wirePosition() {
+		
 		switch (this.superLayerId) {
-			case 0:
+			case 1:
```
same comment about hardcoded numbers

whit2333 replied:
https://github.com/JeffersonLab/coatjava/issues/400

raffaelladevita commented on reconstruction/alert/src/main/java/org/jlab/rec/service/AHDCEngine.java:
```diff
@@ -69,25 +69,26 @@ public boolean processDataEvent(DataEvent event) {
 			ArrayList<Hit>     AHDC_Hits     = hitRead.get_AHDCHits();
 			ArrayList<TrueHit> TrueAHDC_Hits = hitRead.get_TrueAHDCHits();
-
+			//System.out.println("AHDC_Hits size " + AHDC_Hits.size());
```
remove commented out text

raffaelladevita commented on reconstruction/alert/src/main/java/org/jlab/rec/ahdc/KalmanFilter/Hit.java:
This class has a lot of overlap with the ahdc.Hit.Hit class. It should be either merged or defined by extending the other class

whit2333 replied:
https://github.com/JeffersonLab/coatjava/issues/401

raffaelladevita commented on reconstruction/alert/src/main/java/org/jlab/rec/ahdc/KalmanFilter/Hit.java:
From previous studies, EJML has better performance than the apache library and should be considered

whit2333 replied:
https://github.com/JeffersonLab/coatjava/issues/402

raffaelladevita commented on common-tools/clas-detector/src/main/java/org/jlab/detector/decode/CLASDecoder4.java:
```diff
@@ -244,6 +246,26 @@ public List<DetectorDataDgtz>  getEntriesSCALER(DetectorType type,
     }
 
+    public void extractPulses(Event event) {
+        mode3.update(6, null, event, schemaFactory, "BMT::wf", "BMT::adc");
```
The new pulse extraction should be applied to AHDC. BMT here was just used to test the code on existing data and to provide an example.
Same for the other changes below

raffaelladevita commented on etc/bankdefs/hipo4/data.json:
comments in the ATOF::add bank should be updated to be consistent with sector/layer starting from 1

whit2333 replied:
@raffaelladevita, is there a technical reason for this? Does this break something?
 We purposefully did it this way to make the physical detector mapping more direct.

raffaelladevita commented on etc/bankdefs/hipo4/data.json:
```diff
@@ -41,7 +28,27 @@
       { "name":"integral"  , "type":"I", "info":"ADC integral (sum over the pulse)"},
-      { "name":"timestamp" , "type":"L", "info":"timestamp"}
+      { "name":"timestamp" , "type":"L", "info":"timestamp"},
+      { "name":"id"        , "type":"S", "info":"link to BMT::wf"}
```
As mentioned above, these changes should be applied to AHDC

raffaelladevita commented on reconstruction/alert/src/main/java/org/jlab/rec/ahdc/PreCluster/PreClusterFinder.java:
```diff
@@ -23,21 +23,21 @@ private void fill_list(List<Hit> AHDC_hits, ArrayList<Hit> sxlx, int super_layer
 	public void findPreCluster(List<Hit> AHDC_hits) {
 		ArrayList<Hit> s0l0 = new ArrayList<Hit>();
-		fill_list(AHDC_hits, s0l0, 0, 0);
+		fill_list(AHDC_hits, s0l0, 1, 1);
```
The List of Lists should be filled all at once to avoid looping 12 times through the hits

whit2333 replied:
I think this is was updated.

raffaelladevita commented on common-tools/clas-detector/src/main/java/org/jlab/detector/pulse/Pulse.java:
```diff
@@ -0,0 +1,46 @@
+    public int id;
+
+    public float adcMax;
+    public float timeRiseCFA;
```
Since this class is for general use, members' names should be more generic removing the CFA or CFD suffix.

Also the pulse "rise" and "fall" are usually referred to as leading and trailing edge that could be better names

whit2333 replied:
Updated.

raffaelladevita commented on common-tools/clas-detector/src/main/java/org/jlab/detector/pulse/ModeAHDC.java:
```diff
@@ -0,0 +1,243 @@
+	public long timeStamp;
+	public float fineTimeStampResolution;
+	public static final short ADC_LIMIT = 4095; // 2^12-1
+	public float amplitudeFractionCFA;
```
All the pulse specific variables should be local variables in the extract method

baltzell replied:
yes, good catch, critical for thread safety

raffaelladevita committed:
bug fixes

efuchey committed:
Duplicated the measurement matrices functions: one in cylindrical system, one to minimize track distance.
Added functions to calculate phi for hit as a function of z and to calculate the hit "sign" i.e. the side of the wire the track passes.

efuchey committed:
For pull request only: commented out the line in AHDCEngine that gathers all hits to make a track candidate and uncommented the track finding code.

mpaolone committed:
Merge pull request #22 from efuchey/development

Kalman Filter fixes: constants and parameters

baltzell commented:
And the creation of BMT::wf (and conversion to BMT::adc) example in the decoder was intended only as a working example, to be replaced with AHDC.

baltzell committed:
Merge branch 'development' into development

baltzell committed:
fix bad merge conflict

baltzell commented:
Ok, after [this is merged](https://github.com/mpaolone/coatjava/pull/23), then the Mode3 class in the decoder needs to be switched to the one AHDC wants to use.

N-Plx committed:
Updated atof geometry. The index for sector/layer/component matches the actual atof. The detector is oriented as the actual atof.

N-Plx committed:
Merge pull request #389 from JeffersonLab/387-update-atof-geometry-service

Updated atof geometry. The index for sector/layer/component matches t…

mpaolone committed:
Merge pull request #23 from baltzell/development

fix bad merge conflict

mpaolone committed:
changed ModeAHDC to be dumber and threadsafe, changed PreClusterFinder to not loop 12 more times than needed

baltzell committed:
AHDC::wf must be defined

baltzell committed:
actually use the new ahdc pulse extractor

dependabot[bot] committed:
build(deps): bump org.apache.commons:commons-text from 1.12.0 to 1.13.0

Bumps org.apache.commons:commons-text from 1.12.0 to 1.13.0.

---
updated-dependencies:
- dependency-name: org.apache.commons:commons-text
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

raffaelladevita committed:
changed bg-merger default for detector list to be ALL

Commit:
Add AI for ALERT trackfinding

Use DeepJavaLibrary to load a PyTorch model.
Create a new bank AHDC_AI::Prediction.

mpaolone committed:
Merge pull request #24 from baltzell/development

Actually use the new waveform conversion approach for AHDC

whit2333 commented on common-tools/clas-geometry/src/main/java/org/jlab/geom/detector/alert/ATOF/AlertTOFFactory.java:
There should be no changes to ATOF geometry in this PR

baltzell replied:
Yes, due largely to ALERT using a separate repo, but only half of the group actually using it ...

whit2333 reviewed (changes requested):
ATOF geometry should be untouched.

baltzell committed:
Revert "leave internal indexing at zero, output at 1"

This reverts commit 3420bb97c4c247c079dc5479f161f5a799f79848.

c-dilks committed:
ci: test with latest `groovy` version (#394)

mpaolone committed:
Merge pull request #26 from baltzell/development

Revert "leave internal indexing at zero, output at 1"

mpaolone committed:
Merge pull request #25 from mathieuouillon/AI_for_ALERT

Add AI for ALERT trackfinding

baltzell commented:
Who can resolve the ATOF geometry conflict?

raffaelladevita committed:
added new AI-assisted networks for Spring28

baltzell commented:
@whit2333  Can someone help resolve the ATOF geometry conflict?

baltzell committed:
simple truth efficiency calculator (#396)

* update URL

* truth efficiency matrix calculator

* cleanup, add warning printout

* rename, cleanup, add json

* merge scripts

* use an array

* parallelize

* cleanup

* actually loop

* now it works

* use gemc files with truth-matching enabled

* cleanup, run truth efficiency

* bugfix

* cleanup and bugfix

* bugfix, now it works

* avoid unnecessary lookups

* more convenient JSON structure

* cleanup for thread safety

* make adding threadsafe

whit2333 commented:
> @whit2333 Can someone help resolve the ATOF geometry conflict?

What is the conflict?

whit2333 committed:
Merge branch 'big_merge' into dumb_merge

whit2333 reviewed (dismissed):
I think we are OK now.

baltzell commented:
With this merged, mpaolone:development should no longer be part of RG-L's standard git workflow.

whit2333 committed:
Better names for constant fraction variables

- https://github.com/JeffersonLab/coatjava/pull/378#discussion_r1871880348

whit2333 commented:
@raffaelladevita,
I think the review has been completed and I think everything has been addressed or issues created to fix later.

raffaelladevita commented on common-tools/clas-detector/src/main/java/org/jlab/detector/pulse/ModeAHDC.java:
```diff
@@ -2,6 +2,8 @@
 import java.util.List;
 import java.util.ArrayList;
+
+import net.jcip.annotations.GuardedBy;
```
remove unused import

raffaelladevita replied:
#405

raffaelladevita commented on common-tools/clas-detector/src/main/java/org/jlab/detector/pulse/ModeAHDC.java:
```diff
@@ -51,193 +27,232 @@ public class ModeAHDC extends HipoExtractor  {
 	 * Fit the max of the pulse using parabolic fit, this method updates the timeMax and adcMax values
 	 * @param samplingTime time between 2 ADC bins
 	 */
 	private void fitParabolic(float samplingTime) {
```
remove empty method

raffaelladevita replied:
#405

raffaelladevita commented on etc/bankdefs/hipo4/alert.json:
```diff
@@ -159,5 +159,24 @@
+		"item": 30,
+		"info": "Prediction given by AI",
+		"entries": [
+			{"name": "X1", "type": "F", "info": "X1 position of the 1th superprecluster (mm)"},
```
The standard spatial unit in CLAS12 reconstruction is cm.

raffaelladevita replied:
#403

raffaelladevita commented on reconstruction/alert/src/main/java/org/jlab/rec/ahdc/AI/PreClustering.java:
```diff
@@ -0,0 +1,110 @@
+        ArrayList<Hit> s5l1 = fill(AHDC_hits, 5, 1);
+
+        // Sort hits of each layers by phi:
+        s1l1.sort(new Comparator<Hit>() {@Override public int compare(Hit a1, Hit a2) {return Double.compare(a1.getPhi(), a2.getPhi());}});
```
Is this equivalent to an ordering based on wire id? Can this be implemented in the hit class, modifying te compareTo method?

raffaelladevita replied:
#405

raffaelladevita commented on reconstruction/alert/src/main/java/org/jlab/rec/ahdc/AI/PreClustering.java:
```diff
@@ -0,0 +1,110 @@
+                        has_next = false;
+                        for (Hit hit1 : p) {
+                            if (hit1.is_NoUsed() && (hit1.getWireId() == temp.get(temp.size() - 1).getWireId() + 1 || hit1.getWireId() == temp.get(temp.size() - 1).getWireId() - 1)) {
+                                temp.add(hit1);
```
what happens for hits at the min/max of the wireId range?

raffaelladevita replied:
#405

raffaelladevita commented on reconstruction/alert/src/main/java/org/jlab/rec/ahdc/AI/model/model.pt:
Shouldn't this be included as a resource or moved to the etc directory?

raffaelladevita replied:
#404

raffaelladevita commented on reconstruction/alert/src/main/java/org/jlab/rec/ahdc/Track/Track.java:
```diff
@@ -40,6 +40,21 @@ public Track(List<Cluster> clusters) {
+	double p = 150.0;//MeV/c
+	//take first hit.
+	Hit hit = hitslist.get(0);
+	double phi          = Math.atan2(hit.getX(), hit.getY());
```
is phi defined with respect to the vertical direction?

raffaelladevita replied:
#405

whit2333 commented:
@baltzell Can you review this?

raffaelladevita commented on reconstruction/alert/src/main/java/org/jlab/rec/service/AHDCEngine.java:
```diff
@@ -39,11 +60,42 @@ public AHDCEngine() {
+		};
+
+		Criteria<float[], Float> my_model = Criteria.builder().setTypes(float[].class, Float.class)
+				.optModelPath(Paths.get(System.getenv("CLAS12DIR") + "/../reconstruction/alert/src/main/java/org/jlab/rec/ahdc/AI/model/"))
```
See the comment on the network file location. Placing the file with the source code is not a good option because the source code is not included in a standard coatjava installation

raffaelladevita replied:
#404

raffaelladevita reviewed (approved):
All pending changes have been transformed into issues

baltzell commented:
closing just to workaround github bug ...
